libtenzir/builtins/operators/time.cpp

```diff
@@ -0,0 +1,567 @@
+  const auto* result = ::strptime(input, format, &time);
+  if (not result) {
+    diagnostic::error("failed to parse time")
+      .note("strptime error: `{}`", strerror_nonbroken(errno))
+      .hint("input: `{}`, format: `{}`", input, format)
+      .emit(diag);
+    return std::nullopt;
+  }
+  if (result != input + std::strlen(input)) {
+    diagnostic::error("failed to parse time")
+      .note("format string not exhaustive (`{}` not parsed)", result)
+      .hint("input: `{}`, format: `{}`", input, format)
+      .emit(diag);
+    return std::nullopt;
+  }
+  return partial_timestamp::from_tm_with_unset_fields(time, is_unset);
+}
+
+class time_parser final : public plugin_parser {
+public:
+  time_parser() = default;
+
+  explicit time_parser(parser_interface& p) {
+    auto parser = argument_parser{"time", docs};
+    located<std::string> format{};
+    parser.add(format, "<format>");
+    parser.add("--components", components_);
+    parser.add("--strict", strict_);
+    parser.parse(p);
+    format_ = std::move(format.inner);
+  }
+
+  auto name() const -> std::string override {
+    return "time";
+  }
+
+  auto
+  instantiate(generator<chunk_ptr> input, operator_control_plane& ctrl) const
+    -> std::optional<generator<table_slice>> override {
+    (void)input;
+    diagnostic::error("`{}` cannot be used here", name())
+      .emit(ctrl.diagnostics());
+    return {};
+  }
+
+  auto parse_strings(std::shared_ptr<arrow::StringArray> input,
+                     operator_control_plane& ctrl) const
+    -> std::vector<typed_array> override {
+    auto b = series_builder{type{record_type{}}};
+    for (auto&& string : values(string_type{}, *input)) {
+      if (not string) {
+        b.null();
+        continue;
+      }
+      auto& diag = ctrl.diagnostics();
+      auto time
+        = strptime_partial(diag, std::string{*string}.c_str(), format_.c_str());
+      if (not time)
+        return {};
+      if (not strict_) {
+        // If --strict is not set,
+        // we "enrich" `time` with the first second of today.
+        //
+        // In other words, if some of the fields of `time` weren't set by
+        // `strptime_partial`, we'll set them to a reasonable-ish default, that
+        // being today at 00:00:00 UTC.
+        //
+        // If --strict is set:
+        //  - but --components is: the missing fields will just be `null`
+        //  - but --components is NOT: error, as there's not enough information
+        //                             to create a timestamp
+        const bool year_set = time->tm_year.has_value();
+        auto today_beginning = partial_timestamp::today_beginning();
+        if (not time->enrich(today_beginning, diag))
+          return {};
+        TENZIR_ASSERT(time->is_complete());
+        // A special case:
+        // If `strptime_partial` didn't set the year, but the resulting enriched
+        // time would be in the future, subtract 1 from the year.
```

Comment:
I think we should allow the timestamp to be a bit in the future. Other systems do this as well (with varying allowed differences). We should also make sure that this behavior is "smooth".